File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.cpp
#include <string>
#include <vector>
#include <iostream>
using namespace std;

vector<int> I;
vector<vector<int>> al (17,vector<int> {});
int answer = 0;

void dfs(int now ,int sheep, int wolf , vector<int> edges){
    if (I[now] == 0){
        sheep ++;
        answer = max(answer,sheep);
    }else{
        wolf ++;
    }
    
    if(sheep <= wolf){
        return;
    }
    
    vector<int> edges_copy (edges);
    
    for(auto edge : edges){
        vector<int> next_edges (al[edge]);
        
        for (auto x : edges){
            if (x != edge){
                next_edges.push_back(x);
            }
        }
        
        dfs(edge,sheep,wolf,next_edges);
    }
    
}


int solution(vector<int> info, vector<vector<int>> edges) {
    I = info;
    
    for (auto edge : edges){
        al[edge[0]].push_back(edge[1]);
    }
    
    dfs(0,0,0,al[0]);
    
    return answer;
}

File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.py
I = []
al = [[] for _ in range(17)]
answer = 0

def dfs(now,sheep,wolf,edges):
    global al , answer , I

    #print("node: ",now, "-",sheep,wolf,edges)
    if I[now] == 0:
        sheep += 1
        answer = max(answer,sheep)
    else:
        wolf += 1
        
    if sheep <= wolf:
        return
    
    
    for edge in edges:
        next_edges = al[edge] + edges
        dfs(edge,sheep,wolf,[_ for _ in next_edges if _ != edge])
    return
        
def solution(info, edges):
    global I 
    global al, answer
    
    I = info
    
    for edge in edges:
        al[edge[0]].append(edge[1])
        
    dfs(0,0,0,al[0])
    
    return answer